module/os/freebsd/zfs/abd_os.c

```diff
@@ -409,6 +462,50 @@ abd_get_offset_scatter(abd_t *abd, abd_t *sabd, size_t off,
 	return (abd);
 }
 
+#ifdef _KERNEL
+/*
+ * Allocate a scatter ABD structure from user pages.
+ */
+abd_t *
+abd_alloc_from_pages(vm_page_t *pages, unsigned long offset, uint64_t size)
+{
+	VERIFY3U(size, <=, DMU_MAX_ACCESS);
+	ASSERT3U(offset, <, PAGE_SIZE);
+	ASSERT3P(pages, !=, NULL);
+
+	abd_t *abd = abd_alloc_struct(size);
+	abd->abd_flags |= ABD_FLAG_OWNER | ABD_FLAG_FROM_PAGES;
+	abd->abd_size = size;
+
+	if (size < PAGE_SIZE) {
+		/*
+		 * We do not have a full page so we will just use  a linear ABD.
+		 * We have to make sure to take into account the offset though.
+		 * In all other cases our offset will be 0 as we are always
+		 * PAGE_SIZE aligned.
+		 */
+		ASSERT3U(offset + size, <=, PAGE_SIZE);
```

Comment:
What makes you think that if `size < PAGE_SIZE` then we have only one page? Depending on offset it may be two. The assertion looks wrong to me.  We should use its condition above instead.  Same in Linux.

Reply:
So, there is no way we can have two pages here. O_DIRECT is strict on `PAGE_SIZE` alignment. So where this originally came form is if a user issues a `PAGE_SIZE` buffer but the record size is set below `PAGE_SIZE`. Let me know though if you think that the alignment checks could lead to two pages in this case. I did the check though to just be `if ((offset + size) <= PAGE_SIZE)`. The main reason for this is I neglected to think about what if the user just had a `PAGE_SIZE` buffer. We might as well make it a linear page as well.

Reply:
> O_DIRECT is strict on PAGE_SIZE alignment.

It may be strict, but this function seems abstract enough and receives the offset argument.  I see it is currently able to handle non-zero sub-page offsets in MULTI_CHUNK case.  I think you should add offset to the virtual address of the buffer in linear case too.

Reply:
Maybe I am missing something here, but for the multi chunk case it is always asserted that the `offset` is `0`. This is due to the fact of the strict `PAGE_SIZE` alignment checks. There should never be a case where 2 pages are part of the first case. Let me know though if I am just not understanding the issue here.

Reply:
I see now that you have there ASSERT0(abd_offset) in MULTI_CHUNK, after assigned it first.  But I am not sure why do you need that limitation if functions like abd_get_offset_scatter() increment abd_offset to get a fraction of the buffer.  Why  abd_alloc_from_pages() taking offset argument for some reason and asserting first `ASSERT3U(offset, <, PAGE_SIZE)` can not allow it non-zero later?  To me increment of virtual address by offset parameter would be a proper LINEAR_PAGE alternative to abd_offset of MULTI_CHUNK case.

Reply:
So, I believe I am doing exactly that for `LINEAR_PAGE` case already. However, I might just be missing what you are asking to be changed here at this point. Feel free to share a code snippet of what you would like changed, and I will update the code.

Reply:
To me it looks like you are doing it for FreeBSD but not for Linux.

Reply:
Ah okay. So `sg_virt()` gets the offset automatically in the page through `sg->offset`. So, there is no reason to update the Linux code. Let me know if you see the Linux kernel scatter list code differently though.

Reply:
Ah. If you say so. I just don't know that Linux KPI.